plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java

```diff
@@ -144,20 +153,107 @@ public void processResponseAsync(
+                    rewriteResponseDocuments(oneHitResponse, oneHitListener);
+                }
+            }
+
         } catch (Exception e) {
             if (ignoreFailure) {
                 responseListener.onResponse(response);
             } else {
-                responseListener.onFailure(e);
+                responseListener
+                    .onFailure(new OpenSearchStatusException("Failed to process response: " + e.getMessage(), RestStatus.BAD_REQUEST));
             }
         }
     }
 
+    /**
+     * Creates an ActionListener for a single SearchResponse that delegates its
+     * onResponse and onFailure callbacks to a GroupedActionListener.
+     *
+     * @param combineResponseListener The GroupedActionListener to which the
+     *                                 onResponse and onFailure callbacks will be
+     *                                 delegated.
+     * @return An ActionListener that delegates its callbacks to the provided
+     *         GroupedActionListener.
+     */
+    private static ActionListener<SearchResponse> getOneHitListener(GroupedActionListener<SearchResponse> combineResponseListener) {
+        ActionListener<SearchResponse> oneHitListener = new ActionListener<>() {
+            @Override
+            public void onResponse(SearchResponse response) {
+                combineResponseListener.onResponse(response);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                combineResponseListener.onFailure(e);
+            }
+        };
+        return oneHitListener;
+    }
+
+    /**
+     * Creates a GroupedActionListener that combines the SearchResponses from individual hits
+     * and constructs a new SearchResponse with the combined hits.
+     *
+     * @param response         The original SearchResponse containing the hits to be processed.
+     * @param responseListener The ActionListener to be notified with the combined SearchResponse.
+     * @param hits             The array of SearchHits to be processed.
+     * @return A GroupedActionListener that combines the SearchResponses and constructs a new SearchResponse.
+     */
+    private GroupedActionListener<SearchResponse> getCombineResponseGroupedActionListener(
+        SearchResponse response,
+        ActionListener<SearchResponse> responseListener,
+        SearchHit[] hits
+    ) {
+        GroupedActionListener<SearchResponse> combineResponseListener = new GroupedActionListener<>(new ActionListener<>() {
+            @Override
+            public void onResponse(Collection<SearchResponse> responseMapCollection) {
+                SearchHit[] combinedHits = new SearchHit[hits.length];
+                int i = 0;
+                for (SearchResponse OneHitResponseAfterInference : responseMapCollection) {
+                    SearchHit[] hitsAfterInference = OneHitResponseAfterInference.getHits().getHits();
```

Comment:
This code assumes that `hitsAfterInference` always has at least one element (`hitsAfterInference[0]`). If any `SearchResponse` in `responseMapCollection` has no hits (i.e., `hitsAfterInference` is empty), this would throw an `ArrayIndexOutOfBoundsException`.

Reply:
when we made the oneHitResponse, at least it will contain one hit.